.github/workflows/publish.yml

```diff
@@ -0,0 +1,87 @@
+name: Publish Docker image
+on:
+# Will only run when release is published. 
+  release:
+    types: 
+      - created
+  workflow_dispatch:
+
+jobs:
+  publish-images:
+    runs-on: ubuntu-20.04
+    steps:
+      # Set fetch-depth: 0 to fetch commit history and tags for use in version calculation
+      - name: Check out code
+        uses: actions/checkout@v2.3.4
+        with:
+          fetch-depth: 0
+
+      - name: create checksum file
+        run: find . -type f -name "*.gradle*" -o -name "gradle-wrapper*" | xargs shasum | sort > checksum.txt && cat checksum.txt
+
+      - name: Cache packages
+        id: cache-packages
+        uses: actions/cache@v2
+        with:
+          path: ~/.gradle
+          key: gradle-packages-${{ runner.os }}-${{ hashFiles('**/checksum.txt') }}
+
+      - name: download dependencies
+        run: ./gradlew downloadDependencies
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Publish docker image
+        run: ./gradlew publish dockerPushImages
+
+  validate-helm-charts:
+    runs-on: ubuntu-20.04
+    container: hypertrace/helm-gcs-packager:0.3.0
+     # Set fetch-depth: 0 to fetch commit history and tags for use in version calculation
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v2.3.4
+        with:
+          fetch-depth: 0
+
+      - name: validate charts
+        run: |
+          helm dependency update ./helm/
+          helm lint --strict ./helm/
+          helm template ./helm/
+
+  publish-helm-charts:
+    runs-on: ubuntu-20.04
+    container: hypertrace/helm-gcs-packager:0.3.0
+    needs: publish-images, validate-helm-charts
+    steps:
+       # Set fetch-depth: 0 to fetch commit history and tags for use in version calculation
+      - name: Checkout Repository
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}       
+   
+      - name: package and release charts
+        env:
+          HELM_GCS_CREDENTIALS: ${{ secrets.HELM_GCS_CREDENTIALS }}
```

Comment:
We can easily avoid putting secrets in the env here - is it any safer? If there were a script in the below run block, the secret would be compromised. Can we inline them instead to lower the possibly exposures (assuming GH doesn't expose them via the commands being run)?

Reply:
Changed this because of https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets#using-encrypted-secrets-in-a-workflow .
here they mentioned

```
Avoid passing secrets between processes from the command line, whenever possible. Command-line processes may be visible to other users (using the ps command) or captured by security audit events. To help protect secrets, consider using environment variables, STDIN, or other mechanisms supported by the target process.
```

Reply:
This is interesting to me - you're right, they say to do it this way, and they're right, that traditionally you don't want to pass sensitive data via CLI. The part that I'm caught on though, is that this is running in an isolated container (so we don't have to worry about other processes or users seeing via `ps`, at least I think?), but on the other hand, environment variables _are_ available to everything executing in this step so if a script were executing, it could be modified to read the environment variable.

I'm certainly open to be convinced, but my suspicion is that those docs aren't considering the full picture here. Maybe as a middle ground, we use the STDIN approach and pipe the secrets (for example, our old docker login command was `echo $DOCKERHUB_PASSWORD | docker login --username $DOCKERHUB_USERNAME --password-stdin`

Reply:
@JBAhire - you're right, because this is not an external workflow, the secret protection can be a little looser - this is fine.